Refactor Storage.jsm into a class.

Osmose opened the pull request:
This simplifies the implementation of Storage.jsm, as well as usage
outside of the sandbox, at the cost of adding work to the driver to
make the storage instance usable within the sandbox.
SandboxManager.wrapAsync handles the heavy lifting there.

mythmon commented on recipe-client-addon/lib/NormandyDriver.jsm:
```diff
@@ -123,15 +123,19 @@ this.NormandyDriver = function(sandboxManager) {
       }
-      return storage;
+
+      return sandboxManager.cloneInto(storage, {cloneFunctions: true});
```
Instead of modifying and passing the storage object directly, could we make an interface object to pass in?

```js
const storage  = new Storage(prefix);

const interface = {};
for (const method of ["getItem", "setItem", "removeItem", "clear"]) {
  interface[method] = sandboxManager.wrapAsync(storage[method], {
    cloneArguments: true,
    cloneInto: true,
  });
}

return sandboxManager.cloneInto(interface, {cloneFunctions: true});
```

I think this will help avoid exposing unintended functions to the sandbox. I also think that it would be better to not modify the storage object, so any `this.foo` method calls it makes will use unmodified versions of the methods. Would that work?

Osmose commented:
Updated with @mythmon's feedback about using an interface. I added a test to ensure that the methods on the interface object were bound properly; without it, `this.prefix` within the storage methods was null, and the prefix between all storage instances was shared.
